### PR TITLE
Add iOS Sauce Labs tests steps to Buildkite pipeline

### PR DESCRIPTION
**Pull request by mokagio**
This PR adds all the step and automation necessary to build, upload, and run E2E tests on Sauce Labs, but it does not remove the CircleCI steps yet. That will happen in a follow up PR after this new setup run in the wild for a couple days at least.

Unless I missed something, this should replicate the CircleCI configuration 1:1, minus the GitHub comment to start the tests,  more on that later. 

Highlights:

**Block step to manually run optional E2E tests** _[source](https://buildkite.com/automattic/gutenberg-mobile/builds/7205)_

image

**`trunk`, `release/` and Dependabot branches run all the tests** which can be verified by swapping the branch filter definitions in the common nodes at the start of the pipeline file. _[source](https://buildkite.com/automattic/gutenberg-mobile/builds/7159)_

Screenshot 2023-09-11 at 1 58 40 pm

**Slack reporting when tests fail**, verified by adding a failing test to the suite [_source_](https://buildkite.com/automattic/gutenberg-mobile/builds/7181) 

Screenshot 2023-09-07 at 1 05 52 pm

And, additionally, test failure information in the Buildkite build page, [_source_](https://buildkite.com/automattic/gutenberg-mobile/builds/7181):

Screenshot 2023-09-07 at 1 06 01 pm

### About the GitHub comment

Screenshot 2023-09-11 at 2 03 49 pm

We discussed this internally as part of the migration from Peril to [Danger](https://github.com/Automattic/dangermattic/) and  concluded that it's not required. It will be removed in sync with the CircleCI steps removal. Internal ref: paaHJt-5g5-p2#comment-7831

---

PR submission checklist:

- [x] I have considered adding unit tests where possible. – N.A.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary. – N.A.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -128,3 +135,100 @@ steps:
...
+      - *git-cache-plugin
+    artifact_paths:
+      - reports/test-results/ios-test-results.xml
+    agents:
+      queue: mac
+    env: *xcode_agent_env
+    notify:
+      - github_commit_status:
+          context: Test iOS on Device - Canaries
+
```
Comment (mokagio) [inline .buildkite/pipeline.yml:165]: There is quite a bit of duplication below because of the different steps requirement depending on the branch we're in. That is, on `trunk`, `release/` and Dependabot branches, we run _all_ UI test suites, while on other branches we want a block step to manually trigger the full UI test suites.

In the future, I think we should consider using a dynamic pipeline to avoid duplication and make changes simpler. But it seems out of scope for this PR.

File: .buildkite/build-ios.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+echo '--- :node: 2. Load nvm in the current shell'
+export NVM_DIR="$HOME/.nvm"
+mkdir -p "$NVM_DIR"
+[ -s "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" ] && \. "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" --install
+
+echo '--- :node: 3. Install node version from .nvmrc'
+nvm install "$(cat .nvmrc)" && nvm use
+
+echo '--- :node: 4. nmp ci'
+npm ci
```
Comment (mokagio) [inline .buildkite/build-ios.sh:16]: I tried to use the new nvm plugin, but run into a setup issue https://github.com/Automattic/nvm-buildkite-plugin/issues/6 (although it's possible I just misunderstood the plugin target platform)

File: .buildkite/build-ios.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  && popd
+
+echo '--- :saucelabs: Upload app artifact to SauceLabs'
+SAUCE_FILENAME=${BUILDKITE_BRANCH//[\/]/-}
+curl -u "$SAUCE_USERNAME:$SAUCE_ACCESS_KEY" \
+  --location \
+  --request POST 'https://api.us-west-1.saucelabs.com/v1/storage/upload' \
+  --form 'payload=@"./gutenberg/packages/react-native-editor/ios/GutenbergDemo.app.zip"' \
+  --form "name=Gutenberg-$SAUCE_FILENAME.app.zip" \
+  --form 'description="Gutenberg"'
```
Comment (mokagio) [inline .buildkite/build-ios.sh:46]: Just a reminder that we discussed moving off Sauce Labs and agreed it to be something to do in a standalone PR p9ugOq-41A-p2

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -1,26 +1,32 @@
 x-common-params:
   - &gb-mobile-docker-container
     docker#v3.8.0:
-      image: "public.ecr.aws/automattic/gb-mobile-image:latest"
+      image: 'public.ecr.aws/automattic/gb-mobile-image:latest'
```
Comment (mokagio) [inline .buildkite/pipeline.yml:4]: I think VSCode this this for me. Happy to revert if considered noisy. At some point, we might want to establish a standard YAML syntax (or maybe WordPress.org already has one that we can adopt?) but this doesn't seem the time.

File: .buildkite/build-ios.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+echo '--- :node: 2. Load nvm in the current shell'
+export NVM_DIR="$HOME/.nvm"
+mkdir -p "$NVM_DIR"
+[ -s "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" ] && \. "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" --install
+
+echo '--- :node: 3. Install node version from .nvmrc'
+nvm install "$(cat .nvmrc)" && nvm use
+
+echo '--- :node: 4. nmp ci'
```
Comment (dcalhoun) [inline .buildkite/build-ios.sh:15]:
```suggestion
echo '--- :node: 4. npm ci'
```

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (dcalhoun) [reply .buildkite/pipeline.yml:4]: My guess is that Prettier may have done this, if you have the VS Code plugin installed. 

I believe there are a few files in `gutenberg-mobile` and `gutenberg` — e.g. release notes — that do not embrace WordPress coding standards or leverage automated formatting. There is opportunity for us to further embracing those standards and configure the repositories to enforce them via Prettier or eslint.

File: .buildkite/test-ios.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+echo '--- :node: 2. Load nvm in the current shell'
+export NVM_DIR="$HOME/.nvm"
+mkdir -p "$NVM_DIR"
+[ -s "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" ] && \. "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" --install
+
+echo '--- :node: 3. Install node version from .nvmrc'
+nvm install "$(cat .nvmrc)" && nvm use
+
+echo '--- :node: 4. nmp ci (for E2E testing)'
```
Comment (dcalhoun) [inline .buildkite/test-ios.sh]:
```suggestion
echo '--- :node: 4. npm ci (for E2E testing)'
```

**Review by dcalhoun — APPROVED** [review]
🚀 This looks good to me. I left a couple of comments regarding typos. Feel free to await additional feedback, if you'd like.

Thank you completing this work!

File: .buildkite/build-ios.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+echo '--- :node: 1. Install nvm'
+brew install nvm
+
+echo '--- :node: 2. Load nvm in the current shell'
+export NVM_DIR="$HOME/.nvm"
+mkdir -p "$NVM_DIR"
+[ -s "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" ] && \. "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" --install
+
+echo '--- :node: 3. Install node version from .nvmrc'
+nvm install "$(cat .nvmrc)" && nvm use
```
Comment (crazytonyli) [inline .buildkite/build-ios.sh:13]: An alternative to the manual installation above is using [nvm-buildkite-plugin](https://github.com/Automattic/nvm-buildkite-plugin). You can simply add the plugin `automattic/nvm#0.1.0` to the pipeline.yml file and remove the lines above.

File: .buildkite/test-ios.sh [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+echo '--- :node: 1. Install nvm'
+brew install nvm
+
+echo '--- :node: 2. Load nvm in the current shell'
+export NVM_DIR="$HOME/.nvm"
+mkdir -p "$NVM_DIR"
+[ -s "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" ] && \. "$HOMEBREW_PREFIX/opt/nvm/nvm.sh" --install
+
+echo '--- :node: 3. Install node version from .nvmrc'
+nvm install "$(cat .nvmrc)" && nvm use
```
Comment (crazytonyli) [inline .buildkite/test-ios.sh:32]: Same as the above comment.

File: .buildkite/build-ios.sh (hunk 7 above)
Comment (mokagio) [reply .buildkite/build-ios.sh:13]: @crazytonyli I tried to use the new nvm plugin, but run into a setup issue https://github.com/Automattic/nvm-buildkite-plugin/issue/6.

In my previous mention of that issue, I supposed I could have misunderstood the target platform. From your comment, maybe I did something wrong in the setup.

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (mokagio) [reply .buildkite/pipeline.yml:4]: Thanks for the pointer @dcalhoun

File: .buildkite/build-ios.sh (hunk 7 above)
Comment (mokagio) [reply .buildkite/build-ios.sh:13]: Given my current understanding of there being an issue in adopting the plugin, I decided to enable auto-merge. If it turns out I was wrong and it's just a matter of me not having configured it properly, then I'll happily followup with a PR to remove the manual Node setup

Comment (crazytonyli) [reply .buildkite/build-ios.sh:13]: Thanks for reporting the issues. I can have a look at them later. 👍 

> If it turns out I was wrong and it's just a matter of me not having configured it properly

No. Those issues should definitely be solved on the plugin side.
